### PR TITLE
Port middleware connectors to the orchestrator

### PR DESCRIPTION
**Pull request by jdisanti**
This PR ports all the connectors from the `aws-smithy-client` crate into `aws-smithy-runtime` implementing the new `HttpConnector` trait. The old connectors are left in place for now, and follow up PRs will remove them as well as revise the generated configs to take `HttpConnector` impls rather than `DynConnector`.

----

_By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice._

### CODE REVIEW

**Review by ysaito1001 — APPROVED** [review]

**Review by Velfi — APPROVED** [review]
Thanks for converting all of these.

File: CHANGELOG.next.toml [hunk 1]
```diff
@@ -102,3 +102,27 @@ message = "Required members with @contextParam are now treated as client-side re
...
+author = "jdisanti"
+
+[[smithy-rs]]
+message = "DVR's RecordingConnection and ReplayingConnection were renamed to RecordingConnector and ReplayingConnector respectively."
+references = ["smithy-rs#2970"]
+meta = { "breaking" = true, "tada" = false, "bug" = false, "target" = "client" }
+author = "jdisanti"
+
+[[smithy-rs]]
+message = "TestConnection was renamed to EventConnector."
```
[the remark this answers is not in the document]
Comment (jdisanti) [reply CHANGELOG.next.toml:125]: Good call. Added in 17a518c83cfeccf1705669b4dcc4e7f104caeeed

File: CHANGELOG.next.toml [hunk 2]
```diff
@@ -102,3 +102,27 @@ message = "Required members with @contextParam are now treated as client-side re
...
+references = ["smithy-rs#2970"]
+meta = { "breaking" = true, "tada" = false, "bug" = false, "target" = "client" }
+author = "jdisanti"
```
[the remark this answers is not in the document]
Comment (jdisanti) [reply CHANGELOG.next.toml:128]: Talked about in 17a518c83cfeccf1705669b4dcc4e7f104caeeed

File: rust-runtime/aws-smithy-runtime-api/src/client/connectors.rs [hunk 3]
```diff
@@ -3,10 +3,19 @@
...
+use aws_smithy_http::result::ConnectorError;
 use std::fmt;
+use std::future::Future as StdFuture;
+use std::pin::Pin;
 use std::sync::Arc;
 
+/// Boxed future used by [`HttpConnectorFuture`].
+pub type BoxFuture = Pin<Box<dyn StdFuture<Output = Result<HttpResponse, ConnectorError>> + Send>>;
+/// Future for [`HttpConnector::call`].
+pub type HttpConnectorFuture = NowOrLater<Result<HttpResponse, ConnectorError>, BoxFuture>;
```
[the remark this answers is not in the document]
Comment (jdisanti) [reply rust-runtime/aws-smithy-runtime-api/src/client/connectors.rs]: Introduced in 8eb99238c7ec5abe5848183592f2fbaf0fece049

File: rust-runtime/aws-smithy-runtime-api/src/client/runtime_plugin.rs [hunk 4]
```diff
@@ -342,29 +342,29 @@ mod tests {
...
                         .body(SdkBody::empty())
                         .unwrap())
-                })
+                }))
             }
         }
 
         // CN2, the outer connector, calls the inner connector and adds the `rp2` header to the response
         #[derive(Debug)]
         struct CN2(SharedHttpConnector);
```
[the remark this answers is not in the document]
Comment (jdisanti) [reply rust-runtime/aws-smithy-runtime-api/src/client/runtime_plugin.rs]: Renamed in 8eb99238c7ec5abe5848183592f2fbaf0fece049

**Pull request merged**